topcoat-data/topcoat-public
# Added front-end modify columns code, including update to TTable and TSelectColumns

[Pitch](https://www.notion.so/snyk/Table-support-for-modifying-columns-54daf4c649894ff8904e11a175e03312)

Demo: https://user-images.githubusercontent.com/78518576/183450871-294edc61-68ce-4c96-9ec3-525ccb393973.mov


### To test: 

1. in the topcoat IDE check out the modify_columns branch of snyk-insights
2. create a library component called “TestCsvExport.vue” and copy the contents of [TCsvExport](https://github.com/topcoat-data/topcoat-public/blob/851ed2a2f6c9cccd495599c248bd4cad2ec9e779/library/components/TCsvExport.vue) into it.
3. create a visualization called “TSelectColumns.vue” and copy the contents from the [TSelectColumns.vue](https://github.com/topcoat-data/topcoat-public/blob/851ed2a2f6c9cccd495599c248bd4cad2ec9e779/visualizations/frontend/other/TSelectColumns.vue) component in this branch.
4. create a visualization called “TestTable.vue”
- copy the contents of TTable
- update TCsvExport -> TestCsvExport
- update the references to “TTable” in issue_detail.html to “TestTable”
- update the references to “TTable” in issue_pdf_template to “TestTable”
5. Talk to me when it doesn’t work and I’ll figure out what step(s) I missed :joy:

### Usage
There is a new prop on the table. that accepts an array of column configurations in the following format:
:modifiableColumns="[
                {
                    displayColumn: 'SCORE',
                    layerColumns: ['PRIORITY_SCORE'],
                    displayByDefault: true,
                },
                {
                    displayColumn: 'CVE',
                    layerColumns: ['CVE'],
                    displayByDefault: true,
                },
            ]"
Where displayColumn matches the header of the column, layerColumns are the columns(s) in the layer file that the display column relies on, and an optional displayByDefault.

Additionally in the table's layer, each sql column that can be filtered needs to be wrapped in a check:
{% if filter('column_list') | selected_column('PROBLEM_TITLE') %}
	PROBLEM_TITLE,
{% endif %}

Optionally, if there is a join to retrieve columns that won't be used, that join can be averted:
{% if filter('column_list') %}
    {% if filter('column_list') | selected_column(['project_id', 'project_name']) %}
        left join "STITCH_INGESTION"."POSTGRESQL__PROD__SNOWFLAKE"."PROJECTS" p
        on i.project_id = p.id
    {% endif %}
{% else %}
    left join "STITCH_INGESTION"."POSTGRESQL__PROD__SNOWFLAKE"."PROJECTS" p
    on i.project_id = p.id
{% endif %}
